ibc-apps/ics20-transfer/types/src/denom.rs

```diff
@@ -386,18 +401,86 @@ mod tests {
             PrefixedDenom::from_str("transfer/channel-0/transfer/channel-1/uatom").is_ok(),
             "valid multiple trace info"
         );
-        assert!(
-            PrefixedDenom::from_str("(transfer)/channel-0/uatom").is_err(),
-            "invalid port"
-        );
-        assert!(
-            PrefixedDenom::from_str("transfer/(channel-0)/uatom").is_err(),
-            "invalid channel"
-        );
+
+        // the followings are valid denom according to `ibc-go`
+        // https://github.com/cosmos/ibc-go/blob/e2ad31975f2ede592912b86346b5ebf055c9e05f/modules/apps/transfer/types/trace_test.go#L17-L38
+        PrefixedDenom::from_str("/uatom").expect("no error");
+        PrefixedDenom::from_str("//uatom").expect("no error");
+        PrefixedDenom::from_str("transfer/").expect("no error");
+        PrefixedDenom::from_str("transfer/atom").expect("no error");
+        PrefixedDenom::from_str("(transfer)/channel-0/uatom").expect("no error");
+        PrefixedDenom::from_str("transfer/(channel-0)/uatom").expect("no error");
+
+        Ok(())
+    }
+
+    #[rstest]
+    #[case(
+        "transfer/channel-75",
+        "factory/stars16da2uus9zrsy83h23ur42v3lglg5rmyrpqnju4/dust"
+    )]
+    #[case(
+        "transfer/channel-75/transfer/channel-123/transfer/channel-1023/transfer/channel-0",
+        "factory/stars16da2uus9zrsy83h23ur42v3lglg5rmyrpqnju4/dust"
+    )]
+    #[case(
+        "transfer/channel-75/transfer/channel-123/transfer/channel-1023/transfer/channel-0",
+        "//////////////////////dust"
+    )]
+    // https://github.com/cosmos/ibc-go/blob/e2ad31975f2ede592912b86346b5ebf055c9e05f/modules/apps/transfer/types/trace_test.go#L17-L38
+    #[case("", "uatom")]
+    #[case("", "uatom/")]
+    #[case("", "gamm/pool/1")]
+    #[case("", "gamm//pool//1")]
+    #[case("transfer/channel-1", "uatom")]
+    #[case("customtransfer/channel-1", "uatom")]
+    #[case("transfer/channel-1", "uatom/")]
+    #[case(
+        "transfer/channel-1",
+        "erc20/0x85bcBCd7e79Ec36f4fBBDc54F90C643d921151AA"
+    )]
+    #[case("transfer/channel-1", "gamm/pool/1")]
+    #[case("transfer/channel-1", "gamm//pool//1")]
+    #[case("transfer/channel-1/transfer/channel-2", "uatom")]
+    #[case("customtransfer/channel-1/alternativetransfer/channel-2", "uatom")]
```

Comment:
Is it a good idea to distinguish multi-hop scenarios, and check the `TracePath`s order?

Reply:
I added a new test for this.

baedfb7

Reply:
Awesome. Thanks!